File: examples/dataflow-csvio/src/test/java/com/google/example/csvio/TestHelpers.java
/*
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package com.google.example.csvio;

import com.google.example.csvio.SortContextualHeadersAndRows.RecordWithMetadataHelper;
import com.google.gson.Gson;
import java.util.Arrays;
import org.apache.beam.sdk.Pipeline;
import org.apache.beam.sdk.io.contextualtextio.RecordWithMetadata;
import org.apache.beam.sdk.io.fs.ResolveOptions;
import org.apache.beam.sdk.io.fs.ResourceId;
import org.apache.beam.sdk.options.PipelineOptions;
import org.apache.beam.sdk.options.PipelineOptions.CheckEnabled;
import org.apache.beam.sdk.options.PipelineOptionsFactory;
import org.apache.beam.sdk.values.Row;
import org.checkerframework.checker.nullness.qual.Nullable;

class TestHelpers {

  static Pipeline createTestPipeline() {
    PipelineOptions opts = PipelineOptionsFactory.create();
    opts.setStableUniqueNames(CheckEnabled.OFF);
    return Pipeline.create(opts);
  }

  static Row rowFrom(String resourceId, String value, Long recordNum) {
    return Row.withSchema(RecordWithMetadata.getSchema())
        .withFieldValue(RecordWithMetadata.RANGE_OFFSET, 0L)
        .withFieldValue(RecordWithMetadata.RECORD_NUM, recordNum)
        .withFieldValue(RecordWithMetadata.RECORD_OFFSET, 0L)
        .withFieldValue(RecordWithMetadata.RECORD_NUM_IN_OFFSET, 0L)
        .withFieldValue(RecordWithMetadata.RESOURCE_ID, new MockResourceId(resourceId, false))
        .withFieldValue(RecordWithMetadata.VALUE, value)
        .build();
  }

  static ContextualCSVRecord contextualCSVRecordFrom(
      String resourceId, Long recordNum, String header, String record) {
    Row row = rowFrom(resourceId, record, recordNum);
    return ContextualCSVRecord.builder()
        .setResourceId(RecordWithMetadataHelper.getResourceId(row))
        .setLineNumber(recordNum)
        .setHeader(header)
        .setRecord(record)
        .build();
  }

  static Row errorFrom(ContextualCSVRecord record, String message) {
    Gson gson = new Gson();
    return Row.withSchema(CSVIO.ERROR_SCHEMA)
        .withFieldValue(CSVIO.ERROR_FIELD.getName(), message)
        .withFieldValue(CSVIO.REFERENCE_FIELD.getName(), gson.toJson(record))
        .build();
  }

  static class MockResourceId implements ResourceId {

    private static final String PATH_DELIMITER = "/";

    private final String parent;
    private final String name;
    private final boolean isDirectory;

    private MockResourceId(String value, boolean isDirectory) {
      this.isDirectory = isDirectory;
      String[] tokens = value.split(PATH_DELIMITER);
      if (tokens.length <= 1) {
        parent = "";
        name = value;
        return;
      }
      parent = String.join(PATH_DELIMITER, Arrays.asList(tokens).subList(0, tokens.length - 1));
      name = tokens[tokens.length - 1];
    }

    @Override
    public String toString() {
      return parent + PATH_DELIMITER + name;
    }

    @Override
    public ResourceId resolve(String other, ResolveOptions resolveOptions) {
      return null;
    }

    @Override
    public ResourceId getCurrentDirectory() {
      return new MockResourceId(parent, true);
    }

    @Override
    public String getScheme() {
      return "file";
    }

    @Override
    public @Nullable String getFilename() {
      return this.name;
    }

    @Override
    public boolean isDirectory() {
      return this.isDirectory;
    }
  }
}

File: examples/gemini-live-multimodal-avatar/backend/internal/service/test_helpers.go
// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package service

import (
	"context"
)

// MockStore is a shared mock implementation of the internal scenarioStore interface.
// It is exported so it can be used by tests in other packages (e.g., handler_test).
type MockStore struct {
	Scenarios       map[string]map[string]interface{}
	CustomScenarios map[string]map[string]interface{}

	// Error simulation
	GetScenarioErr     error
	GetAllScenariosErr error
	UpdateScenarioErr  error
}

func (m *MockStore) GetScenario(ctx context.Context, collection, id string) (map[string]interface{}, error) {
	if m.GetScenarioErr != nil {
		return nil, m.GetScenarioErr
	}

	col := m.Scenarios
	if collection == "custom_scenarios" {
		col = m.CustomScenarios
	}

	s, ok := col[id]
	if !ok {
		return nil, context.DeadlineExceeded
	}
	return s, nil
}

func (m *MockStore) GetAllScenarios(ctx context.Context, collection string) ([]map[string]interface{}, error) {
	if m.GetAllScenariosErr != nil {
		return nil, m.GetAllScenariosErr
	}

	col := m.Scenarios
	if collection == "custom_scenarios" {
		col = m.CustomScenarios
	}
	var list []map[string]interface{}
	for _, s := range col {
		list = append(list, s)
	}
	return list, nil
}

func (m *MockStore) GetScenariosBySession(ctx context.Context, collection, sessionID string) ([]map[string]interface{}, error) {
	if m.GetAllScenariosErr != nil {
		return nil, m.GetAllScenariosErr
	}

	col := m.CustomScenarios
	var list []map[string]interface{}
	for _, s := range col {
		if sid, ok := s["session_id"].(string); ok && sid == sessionID {
			list = append(list, s)
		}
	}
	return list, nil
}

func (m *MockStore) UpdateScenario(ctx context.Context, collection, id string, scenario map[string]interface{}) error {
	if m.UpdateScenarioErr != nil {
		return m.UpdateScenarioErr
	}

	if collection == "custom_scenarios" {
		if m.CustomScenarios == nil {
			m.CustomScenarios = make(map[string]map[string]interface{})
		}
		m.CustomScenarios[id] = scenario
	} else {
		if m.Scenarios == nil {
			m.Scenarios = make(map[string]map[string]interface{})
		}
		m.Scenarios[id] = scenario
	}
	return nil
}

func (m *MockStore) Close() error {
	return nil
}
